aspnet/fundamentals/caching.rst

```diff
@@ -1,8 +1,157 @@
+
+.. literalinclude:: caching/sample/src/CachingSample/Middleware/GreetingMiddleware.cs
+	:linenos:
+	:lines: 19-28
+	:dedent: 8
+	:emphasize-lines: 2,7
+	
+Reading and Writing to a Memory Cache
+-------------------------------------
+The middleware's ``Invoke`` method uses caching to avoid having to reproduce the same result again and again when the inputs and underlying data haven't changed. A common example of this kind of caching is data-driven navigation menus, which rarely change but are frequently ready for display within an application. Caching results that do not vary often but which are requested frequently can greatly improve performance by reducing round trips to out of process data stores or unnecessary computation.
+
+There are two approaches to attempting to read entries from the cache: ``Get`` and ``TryGet``. ``Get`` will return the value if it exists, but otherwise returns ``null``. ``TryGet`` will assign the cached value to an ``out`` parameter and return true if the entry exists, otherwise it returns false.
+
+Writing to the cache is done via the ``Set`` method, which accepts the key to use to look up the value, the value to be cached, and a set of ``MemoryCacheEntryOptions``. These options allow you to specify absolute or sliding time-based cache expiration, caching priority, callbacks, and dependencies.
+
+In the ``GreetingMiddleware`` example, greetings are cached for one minute. This is implemented using the ``SetAbsoluteExpiration`` method on ``MemoryCacheOptions`` as shown here:
+
+.. literalinclude:: caching/sample/src/CachingSample/Middleware/GreetingMiddleware.cs
+	:linenos:
+	:lines: 30-58
+	:dedent: 8
+	:emphasize-lines: 7,10,16-18
+
+In addition to setting an absolute expiration, a sliding expiration can be used to keep frequently requested items in the cache:
+
+.. code-block:: c#
+
+	// keep item in cache as long as it is requested at least
+	// once every 5 minutes
+	new MemoryCacheEntryOptions()
+		.SetSlidingExpiration(TimeSpan.FromMinutes(5))
+
+To avoid having very active cache entries growing too stale, you can combine absolute and sliding expirations:
+
+.. code-block:: c#
+
+	// keep item in cache as long as it is requested at least
+	// once every 5 minutes...
+	// but in any case make sure to refresh it every hour
+	new MemoryCacheEntryOptions()
+		.SetSlidingExpiration(TimeSpan.FromMinutes(5))
+		.SetAbsoluteExpiration(TimeSpan.FromHours(1))
+
+By default, an instance of ``MemoryCache`` will automatically manage the items stored, removing entries when necessary to make room for new entries. You can influence the way cache entries are managed by setting their `CacheItemPriority <https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Extensions.Caching.Abstractions/CacheItemPriority.cs>`_ when adding the item to the cache. For instance, if you have an item you want to keep in the cache unless you explicitly remove it, you would use the ``NeverRemove`` priority option:
```

Comment:
Cache cleanup is triggered by memory pressure in the overall application, not by Sets or the size of the cache.